python-package/xgboost/core.py

```diff
@@ -418,17 +419,33 @@ def __init__(self, data, label=None, missing=None,
         if label is not None:
             if isinstance(label, np.ndarray):
                 self.set_label_npy2d(label)
+            elif isinstance(label, (CUDF, CUDF_COL)):
+                self.set_cudf_info('label', label)
             else:
                 self.set_label(label)
         if weight is not None:
             if isinstance(weight, np.ndarray):
                 self.set_weight_npy2d(weight)
+            elif isinstance(weight, (CUDF, CUDF_COL)):
+                self.set_cudf_info('weight', weight)
             else:
                 self.set_weight(weight)
 
         self.feature_names = feature_names
         self.feature_types = feature_types
 
+    def _init_from_cudf(self, df):
+        """
+        Initialize data from a GPU data frame.
+        """
+        self.handle = ctypes.c_void_p()
+        col_ptrs = [df[col]._column.cffi_view for col in df.columns]
+        col_ptr_arr = CUDF_FFI.new('gdf_column*[]', col_ptrs)
```

Comment:
This line will require cuDF's foreign function's interface be installed. A byproduct of cuDF installation. Do we want to abstract the FFI out of this initialization? What level of dependency are we OK with in Python?

The `cffi_view` will return `df`'s column pointers, and the `new` call will create a new `cudf` object. We could probably replace this logic with the new functions `InitFromCUDF`, etc.